Caused by: java.lang.ClassNotFoundException: clojure/core$pr_str on JDK11

I'm getting an exception with a basic signal handler like
```clojure
(defn add-signal-handler [signal]
  (signal-handler/with-handler signal
    (log/info ::system-signal-handler {:signal signal})))
```

```
2020-02-28 09:39:47.924 ERROR com.nextjournal.exception-handler - {:msg "Uncaught exception on SIGTERM handler", :line 71}
java.lang.ExceptionInInitializerError: null
 at com.nextjournal.journal$add_signal_handler$fn__69757.invoke(journal.clj:107)
 at signal.handler$__GT_handler$fn__69686.invoke(handler.clj:23)
 at signal.handler.proxy$java.lang.Object$SignalHandler$d8c00ec7.handle(Unknown Source)
 at jdk.unsupported/sun.misc.Signal$InternalMiscHandler.handle(Signal.java:198)
 at java.base/jdk.internal.misc.Signal$1.run(Signal.java:220)
 at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: clojure/core$pr_str
 at java.base/java.lang.Class.forName0(Native Method)
 at java.base/java.lang.Class.forName(Class.java:398)
 at clojure.lang.RT.classForName(RT.java:2211)
 at clojure.lang.RT.classForName(RT.java:2220)
 at clojure.lang.LispReader$EvalReader.invoke(LispReader.java:1317)
 at clojure.lang.LispReader$DispatchReader.invoke(LispReader.java:853)
 at clojure.lang.LispReader.read(LispReader.java:285)
 at clojure.lang.LispReader.read(LispReader.java:216)
 at clojure.lang.LispReader.read(LispReader.java:205)
 at clojure.lang.RT.readString(RT.java:1878)
 at clojure.lang.RT.readString(RT.java:1873)
 at com.nextjournal.journal$add_signal_handler$fn__69757$fn__69760.<clinit>(journal.clj:107)
 ... 6 common frames omitted
```

```sh
$ java -version
Picked up _JAVA_OPTIONS: -XX:+UseG1GC
openjdk version "11.0.6" 2020-01-14
OpenJDK Runtime Environment GraalVM CE 20.0.0 (build 11.0.6+9-jvmci-20.0-b02)
OpenJDK 64-Bit Server VM GraalVM CE 20.0.0 (build 11.0.6+9-jvmci-20.0-b02, mixed mode, sharing)
```

Do you have any idea what's going on here?

--- a/src/signal/handler.clj
+++ b/src/signal/handler.clj
@@ -1,5 +1,16 @@
 (ns signal.handler
-  "Signal handling functions")
+  "Signal handling functions"
+  (:import clojure.lang.Compiler
+           clojure.lang.DynamicClassLoader))
+
+(defn- set-compiler-class-loader
+  "Ensures that the class loader in which a handler will
+   execute runs in a context which can load clojure classes."
+  []
+  (let [thread (Thread/currentThread)]
+    (when-not (instance? DynamicClassLoader (.getContextClassLoader thread))
+      (.setContextClassLoader thread (DynamicClassLoader.
+                                      (.getClassLoader Compiler))))))
 
 (defn ^sun.misc.Signal ->signal
   "Convert a keyword to an appropriate Signal instance."
@@ -20,7 +31,9 @@
   "Convert class to signal handler."
   [handler]
   (proxy [sun.misc.SignalHandler] []
-    (handle [sig] (handler (signal->kw sig)))))
+    (handle [sig]
+      (set-compiler-class-loader)
+      (handler (signal->kw sig)))))
 
 (defn on-signal
   "Execute handler when signal is caught"

